File: CustomApplications/VB.Net/GraphicsHowTo/Primitives/TriangleMesh/TriangleMeshWithHoleCodeSnippet.vb
#Region "UsingDirectives"
Imports System.Collections.Generic
Imports System.Drawing
Imports System.IO
Imports AGI.STKGraphics
Imports AGI.STKObjects
Imports AGI.STKUtil
#End Region

Namespace Primitives.TriangleMesh
	Class TriangleMeshWithHoleCodeSnippet
		Inherits CodeSnippet
		Public Sub New()
            MyBase.New("Primitives\TriangleMesh\TriangleMeshWithHoleCodeSnippet.vb")
        End Sub

        Public Overrides Sub Execute(ByVal scene As IAgStkGraphicsScene, ByVal root As AgStkObjectRoot)
            Dim positions As Array = STKUtil.ReadAreaTargetPoints(New AGI.DataPath(AGI.DataPathRoot.Relative, "AreaTargets/LogoBoundary.at").FullPath, root)
            Dim holePositions As Array = STKUtil.ReadAreaTargetPoints(New AGI.DataPath(AGI.DataPathRoot.Relative, "AreaTargets/LogoHole.at").FullPath, root)
            ExecuteSnippet(scene, root, positions, holePositions)
        End Sub

		' Name        
		' Description 
		' Category    
		' References  
		' Namespaces  
		' EID         
        <AGI.CodeSnippets.CodeSnippet( _
            "DrawAFilledPolygonWithHole", _
            "Draw a filled polygon with a hole on the globe", _
            "Graphics | Primitives | Triangle Mesh Primitive", _
            "System", _
            "System", _
            "AgSTKGraphicsLib~IAgStkGraphicsTriangleMeshPrimitive" _
            )> _
        Public Sub ExecuteSnippet(<AGI.CodeSnippets.CodeSnippet.Parameter("Scene", "Current Scene")> ByVal scene As IAgStkGraphicsScene, <AGI.CodeSnippets.CodeSnippet.Parameter("Root", "STK Object Model root")> ByVal root As AgStkObjectRoot, <AGI.CodeSnippets.CodeSnippet.Parameter("positions", "The positions used to compute triangulation")> ByVal positions As Array, <AGI.CodeSnippets.CodeSnippet.Parameter("holePositions", "The positions of the hole")> ByVal holePositions As Array)
            '#Region "CodeSnippet"
            Dim manager As IAgStkGraphicsSceneManager = DirectCast(root.CurrentScenario, IAgScenario).SceneManager

            Dim triangles As IAgStkGraphicsSurfaceTriangulatorResult = manager.Initializers.SurfacePolygonTriangulator.ComputeWithHole(attachID("$planetName$The planet on which the triangle mesh will be placed$", "Earth"), positions, holePositions)

            Dim mesh As IAgStkGraphicsTriangleMeshPrimitive = manager.Initializers.TriangleMeshPrimitive.Initialize()
            mesh.SetTriangulator(DirectCast(triangles, IAgStkGraphicsTriangulatorResult))
            DirectCast(mesh, IAgStkGraphicsPrimitive).Color = attachID("$color$The System.Drawing.Color of the triangle mesh$", Color.Gray)
            DirectCast(mesh, IAgStkGraphicsPrimitive).Translucency = attachID("$translucency$The translucency of the triangle mesh$", 0.5F)
            manager.Primitives.Add(DirectCast(mesh, IAgStkGraphicsPrimitive))

            Dim boundaryLine As IAgStkGraphicsPolylinePrimitive = manager.Initializers.PolylinePrimitive.Initialize()
            Dim boundaryPositionsArray As Array = triangles.BoundaryPositions
            boundaryLine.Set(boundaryPositionsArray)
            DirectCast(boundaryLine, IAgStkGraphicsPrimitive).Color = attachID("$outlineColor$The color of the outline around the area target$", Color.Red)
            boundaryLine.Width = attachID("$outlineWidth$The width of the outline$", 2)
            manager.Primitives.Add(DirectCast(boundaryLine, IAgStkGraphicsPrimitive))

            Dim holeLine As IAgStkGraphicsPolylinePrimitive = manager.Initializers.PolylinePrimitive.Initialize()
            holeLine.Set(holePositions)
            DirectCast(holeLine, IAgStkGraphicsPrimitive).Color = attachID("$holeOutlineColor$The System.Drawing.Color of the hole's outline$", Color.Red)
            holeLine.Width = attachID("$holeOutlineWidth$The width of the hole's outline$", 2)
            manager.Primitives.Add(DirectCast(holeLine, IAgStkGraphicsPrimitive))
            '#End Region

            m_Primitive = DirectCast(mesh, IAgStkGraphicsPrimitive)
            m_BoundaryLine = DirectCast(boundaryLine, IAgStkGraphicsPrimitive)
            m_HoleLine = DirectCast(holeLine, IAgStkGraphicsPrimitive)
            OverlayHelper.AddTextBox("SurfacePolygonTriangulator.Compute has overloads that take the exterior " & vbCrLf & _
                                     "boundary positions as well as positions for an interior hole.", manager)
        End Sub

		Public Overrides Sub View(scene As IAgStkGraphicsScene, root As AgStkObjectRoot)
			ViewHelper.ViewBoundingSphere(scene, root, "Earth", m_Primitive.BoundingSphere)
			scene.Render()
		End Sub

		Public Overrides Sub Remove(scene As IAgStkGraphicsScene, root As AgStkObjectRoot)
			Dim manager As IAgStkGraphicsSceneManager = DirectCast(root.CurrentScenario, IAgScenario).SceneManager
			manager.Primitives.Remove(m_Primitive)
			manager.Primitives.Remove(m_BoundaryLine)
			manager.Primitives.Remove(m_HoleLine)
			m_Primitive = Nothing
			m_BoundaryLine = Nothing
			m_HoleLine = Nothing

			OverlayHelper.RemoveTextBox(manager)
			scene.Render()
		End Sub

		Private m_Primitive As IAgStkGraphicsPrimitive
		Private m_BoundaryLine As IAgStkGraphicsPrimitive
		Private m_HoleLine As IAgStkGraphicsPrimitive
	End Class
End Namespace

File: CustomApplications/CSharp/GraphicsHowTo/Primitives/TriangleMesh/TriangleMeshWithHoleCodeSnippet.cs
#region UsingDirectives
using System.Collections.Generic;
using System.Drawing;
using System.IO;
using AGI.STKGraphics;
using AGI.STKObjects;
using AGI.STKUtil;
using System;
#endregion

namespace GraphicsHowTo.Primitives.TriangleMesh
{
    class TriangleMeshWithHoleCodeSnippet : CodeSnippet
    {
        public TriangleMeshWithHoleCodeSnippet()
            : base(@"Primitives\TriangleMesh\TriangleMeshWithHoleCodeSnippet.cs")
        {
        }

        public override void Execute(IAgStkGraphicsScene scene, AgStkObjectRoot root)
        {
            Array positions = STKUtil.ReadAreaTargetPoints(new AGI.DataPath(AGI.DataPathRoot.Relative, "AreaTargets/LogoBoundary.at").FullPath, root);
            Array holePositions = STKUtil.ReadAreaTargetPoints(new AGI.DataPath(AGI.DataPathRoot.Relative, "AreaTargets/LogoHole.at").FullPath, root);
            Execute(scene, root, positions, holePositions);
        }

        [AGI.CodeSnippets.CodeSnippet(
            /* Name        */ "DrawAFilledPolygonWithHole",
            /* Description */ "Draw a filled polygon with a hole on the globe",
            /* Category    */ "Graphics | Primitives | Triangle Mesh Primitive",
            /* References  */ "System",
            /* Namespaces  */ "System",
            /* EID         */ "AgSTKGraphicsLib~IAgStkGraphicsTriangleMeshPrimitive"
            )]
        public void Execute([AGI.CodeSnippets.CodeSnippet.Parameter("Scene", "Current Scene")] IAgStkGraphicsScene scene, [AGI.CodeSnippets.CodeSnippet.Parameter("Root", "STK Object Model root")] AgStkObjectRoot root, [AGI.CodeSnippets.CodeSnippet.Parameter("positions", "The positions used to compute triangulation")] Array positions, [AGI.CodeSnippets.CodeSnippet.Parameter("holePositions", "The positions of the hole")] Array holePositions)
        {
#region CodeSnippet
            IAgStkGraphicsSceneManager manager = ((IAgScenario)root.CurrentScenario).SceneManager;

            IAgStkGraphicsSurfaceTriangulatorResult triangles = manager.Initializers.SurfacePolygonTriangulator.ComputeWithHole(
                /*$planetName$The planet on which the triangle mesh will be placed$*/"Earth", ref positions, ref holePositions);

            IAgStkGraphicsTriangleMeshPrimitive mesh = manager.Initializers.TriangleMeshPrimitive.Initialize();
            mesh.SetTriangulator((IAgStkGraphicsTriangulatorResult)triangles);
            ((IAgStkGraphicsPrimitive)mesh).Color = /*$color$The System.Drawing.Color of the triangle mesh$*/Color.Gray;
            ((IAgStkGraphicsPrimitive)mesh).Translucency = /*$translucency$The translucency of the triangle mesh$*/0.5f;
            manager.Primitives.Add((IAgStkGraphicsPrimitive)mesh);

            IAgStkGraphicsPolylinePrimitive boundaryLine = manager.Initializers.PolylinePrimitive.Initialize();
            Array boundaryPositionsArray = triangles.BoundaryPositions;
            boundaryLine.Set(ref boundaryPositionsArray);
            ((IAgStkGraphicsPrimitive)boundaryLine).Color = /*$outlineColor$The color of the outline around the area target$*/Color.Red;
            boundaryLine.Width = /*$outlineWidth$The width of the outline$*/2;
            manager.Primitives.Add((IAgStkGraphicsPrimitive)boundaryLine);

            IAgStkGraphicsPolylinePrimitive holeLine = manager.Initializers.PolylinePrimitive.Initialize();
            holeLine.Set(ref holePositions);
            ((IAgStkGraphicsPrimitive)holeLine).Color = /*$holeOutlineColor$The System.Drawing.Color of the hole's outline$*/Color.Red;
            holeLine.Width = /*$holeOutlineWidth$The width of the hole's outline$*/2;
            manager.Primitives.Add((IAgStkGraphicsPrimitive)holeLine);
#endregion

            m_Primitive = (IAgStkGraphicsPrimitive)mesh;
            m_BoundaryLine = (IAgStkGraphicsPrimitive)boundaryLine;
            m_HoleLine = (IAgStkGraphicsPrimitive)holeLine;
            OverlayHelper.AddTextBox(
@"SurfacePolygonTriangulator.Compute has overloads that take the exterior 
boundary positions as well as positions for an interior hole.", manager);
        }

        public override void View(IAgStkGraphicsScene scene, AgStkObjectRoot root)
        {
            ViewHelper.ViewBoundingSphere(scene, root, "Earth", m_Primitive.BoundingSphere);
            scene.Render();
        }

        public override void Remove(IAgStkGraphicsScene scene, AgStkObjectRoot root)
        {
            IAgStkGraphicsSceneManager manager = ((IAgScenario)root.CurrentScenario).SceneManager;
            manager.Primitives.Remove(m_Primitive);
            manager.Primitives.Remove(m_BoundaryLine);
            manager.Primitives.Remove(m_HoleLine);
            m_Primitive = null;
            m_BoundaryLine = null;
            m_HoleLine = null;

            OverlayHelper.RemoveTextBox(manager);
            scene.Render();
        }

        private IAgStkGraphicsPrimitive m_Primitive;
        private IAgStkGraphicsPrimitive m_BoundaryLine;
        private IAgStkGraphicsPrimitive m_HoleLine;
    };
}
